### DISCUSSION TITLE
Trouble with --autoinstall "Failed to retrieve latest version"

### DISCUSSION THREAD

**Discussion by thatguythere98** [markqvist/Reticulum#259] [Help & Questions]
Hello, I am having an issue, I am on a Mac running MacOS Monterrey and a LILYGO® TTGO LoRa32 V2.1_1.6 Version 915Mhz.
The installer tells me "Failed to retrieve latest version information for your board from the default server" and later tells me to check my internet connection, but that is working fine.

I am new to this software and this board. I am not sure what the problem is so I have pasted the contents of my terminal. I saw an older thread with a similar problem but they were on windows.
I'm not sure why the board shows up with two ports but I've tried it with both and I always get the same result.

Here is what I get.

```
Hello!

This guide will help you install the RNode firmware on supported
and homebrew devices. Please connect the device you wish to set
up now. Hit enter when it is connected.

Detected serial ports:
  [1] /dev/cu.usbmodem54FC0443881 (USB Single Serial, 54FC044388)
  [2] /dev/cu.wchusbserial54FC0443881 (USB Single Serial, 54FC044388)

Enter the number of the serial port your device is connected to:
? 1
```
```
Ok, using device on /dev/cu.usbmodem54FC0443881 (USB Single Serial, 54FC044388)

Probing device...
[23:11:27] No answer from device

---------------------------------------------------------------------------
                               Device Selection

It looks like this is a fresh device with no RNode firmware.

What kind of device is this?

[1] A specific kind of RNode
       .
      / \   Select this option if you have an RNode of a specific
       |    type, built from a recipe or bought from a vendor.

[2] Homebrew RNode
       .
      / \   Select this option if you have put toghether an RNode
       |    of your own design, or if you are prototyping one.

[3] LilyGO LoRa32 v2.1 (aka T3 v1.6 / T3 v1.6.1)
[4] LilyGO LoRa32 v2.0
[5] LilyGO T-Beam
[6] Heltec LoRa32 v2
       .
      / \   Select one of these options if you want to easily turn
       |    a supported development board into an RNode.

---------------------------------------------------------------------------

Enter the number that matches your device type:
? 3 
```
```
---------------------------------------------------------------------------
                     LilyGO LoRa32 v2.1 RNode Installer

Important! Using RNode firmware on LoRa32 devices should currently be
considered experimental. It is not intended for production or critical use.
The currently supplied firmware is provided AS-IS as a courtesey to those
who would like to experiment with it. Hit enter to continue.
---------------------------------------------------------------------------


What band is this LoRa32 for?

[1] 433 MHz
[2] 868 MHz
[3] 915 MHz
[4] 923 MHz

? 3
```
```
------------------------------------------------------------------------------
                               Installer Ready

Ok, that should be all the information we need. Please confirm the following
summary before proceeding. In the next step, the device will be flashed and
provisioned, so make sure that you are satisfied with your choices.

Serial port     : /dev/cu.usbmodem54FC0443881
Device type     : LilyGO LoRa32 v2.1 850 - 950 MHz
Platform        : ESP32
Device MCU      : Espressif Systems ESP32
Firmware file   : rnode_firmware_lora32v21.zip

------------------------------------------------------------------------------

Is the above correct? [y/N] y
[23:16:13] Checking firmware file availability...
[23:16:14] 
[23:16:14] WARNING!
[23:16:14] Failed to retrieve latest version information for your board from the default server
[23:16:14] Will retry using the following fallback URL: https://github.com/markqvist/rnode_firmware/releases/latest/download/release.json
[23:16:14] 
[23:16:14] Hit enter if you want to proceed

[23:16:16] Error while trying fallback URL: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)>
[23:16:16] Failed to retrive latest version information for your board.
[23:16:16] Check your internet connection and try again.
[23:16:16] If you don't have Internet access currently, use the --fw-version option to manually specify a version.
[23:16:16] You can also use --extract to copy the firmware from a known-good RNode of the same model.
```

**Comment by markqvist** [comment]
Looks like you have a problem with SSL certificates on your system. It could be as simple as having your clock set to an incorrect date. Can you verify that websites over SSL load correctly?

**Reply by thatguythere98** [reply]
My clock and date is correct, websites load fine on SSL. I have reinstalled python and ran the install certificate command but the problem persists.

**Reply by faragher** [reply]
> My clock and date is correct, websites load fine on SSL. I have reinstalled python and ran the install certificate command but the problem persists.

Do you have problems with curl or wget? There are some underlying problems in certain versions and distros that drive me mad and I never looked into it enough to fix them.

**Reply by thatguythere98** [reply]
Both curl and wget work fine from my testing. Is there another way to install without the autoinstaller? I'm not sure what else to do.

**Reply by faragher** [reply]
Apologies, last one was from e-mail. I don't have a rnode in reach, so I need to reverse engineer from my notes:

Download appropriate firmware from https://github.com/markqvist/RNode_Firmware/releases/tag/1.59 - it seems you want rnode_firmware_lora32v21.zip

While you're there, grab release.json

IIRC, the directory in question is [OS Home]/.config/rnodeconf/update/1.59/ - put the zip in that directory . . . I also extracted it, but I don't know if that made any difference.

I think the command is: 

`rnodeconf --autoinstall --nocheck --fw-version 1.59 `

It should provide a number of error messages if something went wrong. Bear in mind, it *shouldn't* work at this time.

Make a file called rnode_firmware_lora32v21.zip.version - the contents of this file should be a word, a space, and the hash. I think 

`hash e904af2582428a24219ebbf88295561f82058be4a1f6ccc5bdfe68d83d14c315`

is appropriate for this version. Otherwise check release.json.

I apologize I can't do this with hardware in front of me, but hopefully my memory and notes, as well as the newest commit, can get this running for you.

Also, you can extract the firmware from a functioning RNode of identical hardware if you have a known good radio.

**Reply by thatguythere98** [reply]
So I tried doing but I realized I don't have this directory [OS Home]/.config/rnodeconf/update/1.59/, I tried making it but the system wont let me create a folder that starts with "."

I think this may be the root cause of my problem. How can I go about creating this folder?

**Reply by faragher** [reply]
I'm afraid I use Linux and Windows, not Mac. I did find this: https://stackoverflow.com/questions/5891365/mac-os-x-doesnt-allow-to-name-files-starting-with-a-dot-how-do-i-name-the-hta

Short version: it should(?) work in terminal, but it should also work if you show hidden files in the GUI. They say the shortcut is Command+Shift+.

To clarify, the [OS Home] should be ~ or /home/[user]/ on Linux or /users/[user] on Windows, I don't know about Mac.

Between the double detection, the SSL error, and the directory issue, I think the changes from Linux to MacOS is causing trouble. If you get it working, please document the steps for it.

**Reply by thatguythere98** [reply]
I was able to find the correct directory, [OS Home]/.config/rnodeconf/update/1.59/, I didn't know how to see hidden files previously. 

When you say make a file called "rnode_firmware_lora32v21.zip.version" how should I create this file? Is it a text file?

Thank you for your patience by the way, I'm not that great at software. I may just create a linux partition on my windows machine because it seems like it would work smoother that way.

**Reply by faragher** [reply]
I love Linux. If it wasn't for development and gaming, I'd never touch Windows again. (I am well aware of WINE, Proton, etc. Edge cases are an issue for me) I suggest playing with it at some point. If you're familiar with the terminal on Mac you'll be well on your way.

You can use a basic text editor to make the file. Just make sure it's super basic and you save it as a text file with that exact file name and extension. If you have something like nano for the command line or something similar to notepad, that will work just fine.

I've been pretty close to where you are, just on Windows, so I'm happy to help. Plenty of time lately.

**Reply by thatguythere98** [reply]
I was finally able to install the firmware. I did just as you said and also had to install drivers for the CH9102 chip in the board and then it finally worked. 

Thank you so much!

**Reply by faragher** [reply]
Fantastic! Have fun with your new radio.

**Comment by faragher** [comment]
Yes, I had to do it because of some hyper-aggressive network security over
here. It'll be about an hour until I can get to my computer to give you
specifics, but the last batch of commits included comments for just this
operation.

Use the offline cache command line parameter and it'll ask for a version
number. That should give you the directory it's looking for. Put the
firmware in that directory and when it complains about a lack of a hash in
the "version" file, make a new file with the appropriate name, put a random
word in the file, press space, then paste the hash for your chosen firmware
and version from the json file in the online repository. If the hash
doesn't work, it may be a comma instead of a space.

It's a work-around I had to do locally, so I documented it in the error
messages. It's messy, but until a full offline load system is implemented,
it does work.
